Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift

```diff
@@ -0,0 +1,324 @@
+            self.rawValue = rawValue
+        }
+        var debugDescription: String { self.rawValue }
+    }
+
+    /// One of the many `PBXObject` types encountered in the `.pbxproj` file format.
+    /// Represents the root project object.
+    struct PBXProject: PBXObject {
+        private let isa: ISA<Self>
+
+        let mainGroup: ObjectUUID
+        let targets: [ObjectUUID]
+    }
+
+    /// One of the many `PBXObject` types encountered in the `.pbxproj` file format.
+    /// Represents a native target (i.e. a target building an app, app extension, bundle...).
+    /// - note: Does not represent other types of targets like `PBXAggregateTarget`, only native ones.
+    struct PBXNativeTarget: PBXObject {
+        private let isa: ISA<Self>
+
+        let name: String
+        let buildPhases: [ObjectUUID]
+        let productType: ProductType
+
+        enum ProductType: String, Decodable {
+            case app = "com.apple.product-type.application"
+            case appExtension = "com.apple.product-type.app-extension"
+            case unitTest = "com.apple.product-type.bundle.unit-test"
+            case uiTest = "com.apple.product-type.bundle.ui-testing"
+        }
+    }
+
+    /// One of the many `PBXObject` types encountered in the `.pbxproj` file format.
+    /// Represents a "Compile Sources" build phase containing a list of files to compile.
+    /// - note: Does not represent other types of Build Phases that could exist in the project, only "Compile Sources" one
+    struct PBXSourcesBuildPhase: PBXObject {
+        private let isa: ISA<Self>
+
+        let files: [ObjectUUID]
+    }
+
+    /// One of the many `PBXObject` types encountered in the `.pbxproj` file format.
+    /// Represents a single build file in a `PBXSourcesBuildPhase` build phase.
+    struct PBXBuildFile: PBXObject {
+        private let isa: ISA<Self>
+
+        let fileRef: ObjectUUID
+    }
+
+    /// This type is used to indicate what a file reference in the project is actually relative to
+    enum SourceTree: String, Decodable {
+        case absolute = "<absolute>"
+        case group = "<group>"
+        case projectRoot = "SOURCE_ROOT"
+        case buildProductsDir = "BUILT_PRODUCTS_DIR"
+        case devDir = "DEVELOPER_DIR"
+        case sdkDir = "SDKROOT"
+    }
+
+    /// One of the many `PBXObject` types encountered in the `.pbxproj` file format.
+    /// Represents a reference to a file contained in the project tree.
+    struct PBXFileReference: PBXReference {
+        private let isa: ISA<Self>
+
+        let name: String?
+        let path: String?
+        let sourceTree: SourceTree
+    }
+
+    /// One of the many `PBXObject` types encountered in the `.pbxproj` file format.
+    /// Represents a group (aka "folder") contained in the project tree.
+    struct PBXGroup: PBXReference {
+        enum ISA: String, Decodable { case PBXGroup, XCVersionGroup, PBXVariantGroup }
+        private let isa: ISA
```

Comment:
ℹ️ We don't use the `ISA<Self>` trick on this one because we want multiple `isa` values (`"PBXGroup"`, `"XCVersionGroup"`, `"PBXVariantGroup"`) to all be allowed and all decode as a `PBXGroup` instance.

Alternatively, I could have created 3 separate `struct PBXGroup`, `struct XCVersionGroup` and `struct PBXVariantGroup`, each using `ISA<Self>` like the other `PBXObject` types. But that would have required to duplicate the same list of fields 3 times, while in our specific scenario and use case we don't care about any field that would otherwise make those 3 types differ, and only care about their 4 properties listed below for our usage, so it was not worth duplicating each type.

Another route I tried was to use inheritance instead, but that opened its own can of worms (like having to implement a `required init` ourselves unlike when using `struct` where we can rely on the auto-generated one, not being able to use `ISA<Self>` if `Self` (aka the class in question) is not `final`, …), so I decided to keep `struct` and value types all the way and use this trick instead after all.

Reply:
> ℹ️ We don't use the `ISA<Self>` trick on this one because we want multiple `isa` values (`"PBXGroup"`, `"XCVersionGroup"`, `"PBXVariantGroup"`) to all be allowed and all decode as a `PBXGroup` instance.

Do you think it would be beneficial to leave this in the code so that future readers know about it?

```suggestion
        enum ISA: String, Decodable { case PBXGroup, XCVersionGroup, PBXVariantGroup }
        // We don't have a `ISA<Self>` here because we want multiple `isa` values to all be allowed and all decode as a `PBXGroup` instance (`"PBXGroup"`, `"XCVersionGroup"`, `"PBXVariantGroup"`)
        private let isa: ISA
```

---

As for the different approaches... I think this is fine. This is a self contained script so we don't need to have it _perfect_ or worry about extensibility and tradeoffs too much 😄 👍

Reply:
Actually the alternate approach I have in my backback (but still haven't found the time in my crazy MIE schedule to apply and push… so by now we should probably merge the PR as is and I'll do a follow-up PR to apply my new approach afterwards on top of it, I guess) don't use the `ISA<Self>` trick anymore.

Instead with the new approach I found a way to decode the `isa` field and then re-decode the same `KeyedContainer` but this time with the right class (`switch`-ing on the `isa` value to decode it), all in one go, i.e. without having to loop over all the possible `Decodable` classes until one succeeds.

In that approach I have in mind, the `isa` is a `String` and then each `PBXObject` type conforms to a protocol which has a `static func match(isa: String) -> Bool` requirement with a default implementation of `static func match(isa: String) { isa == String(describing: self) }`. Then all my structs representing `PBXObjects` use that default implementation, except `PBXGroup` which provides an explicit implementation doing `["PBXGroup", "PBXVariantGroup", "XCVersionGroup"].contains(isa)`.

I could still provide a comment about this even in the new approach (if I get to push it one day…), but I think the new approach will make it more understandable (not having the reader of the code understand how the `ISA<Self>` trick works) and might thus make the comment less necessary…

Reply:
Comment added via your suggestion in [9835f03](https://github.com/wordpress-mobile/WordPress-iOS/pull/18462/commits/9835f037667cd6209349944e06188924cd0a5a11)